xscen/ensembles.py

```diff
@@ -441,7 +475,85 @@ def generate_weights(
+    if attribute_weights:
+        stationary_weights = {}
+        non_stationary_weights = {}
+        for att, v_att in attribute_weights.items():
+            # Add warning when mismatch between independance_level/experiment_weight and attribute_weights
+            if att != independence_level or (
+                att == "experiment" and not balance_experiments
+            ):
+                if att != "experiment":
+                    warnings.warn(
+                        f"The {att} weights do not match the {independence_level} independance_level"
+                    )
+                else:
+                    warnings.warn(
+                        "Key experiment given in attribute_weights without argument balance_experiments=True"
+                    )
+
+            # Verification
+            if att not in info[k] or any(
+                (info[k][att] is None or len(info[k][att]) == 0) for k in info
+            ):
+                raise ValueError(
+                    f"The {att} attribute is missing from some simulations."
+                )
+            # Split dict and xr.Dataarray weights
+            if isinstance(v_att, xr.DataArray):
+                non_stationary_weights[att] = v_att
+            elif isinstance(v_att, dict):
+                stationary_weights[att] = v_att
+            else:
+                raise ValueError("Attribute_weights should be dict or xr.DataArray.")
+        # Stationary weights (dicts)
+        if stationary_weights:
+            for att, w_dict in stationary_weights.items():
+                for k, v in info.items():
+                    if v[att] not in w_dict and "others" not in w_dict:
+                        raise ValueError(
+                            f"The {att} {v[att]} or others are not in the attribute_weights dict."
+                        )
+                    elif v[att] not in w_dict and "others" in w_dict:
+                        w = w_dict["others"]
+                    elif v[att] in w_dict:
+                        w = w_dict[v[att]]
+                    weights.loc[{"realization": k}] = weights.sel(realization=k) * w
+        # Non-staionary weights (xr.DataArray)
+        if non_stationary_weights:
+            for att, da in non_stationary_weights.items():
+                # verification
+                if att not in da.coords:
+                    raise ValueError(f"{att} is not in the xr.DataArray coords.")
+                ls_coord = list(da.coords)
+                ls_coord.remove(att)
+                if len(ls_coord) > 1:
+                    raise ValueError(
+                        f"The {att} DataArray has more than one coord dimension to apply weights."
+                    )
+                else:
+                    coord = ls_coord[0]
+                # coord which will be used for broadcasting
+                if coord not in weights.coords:
+                    weights = weights.expand_dims({coord: da[coord].values})
```

Comment:
I understanding doing this here to simplify the multiplication of matrices, but I think that we want to re-collapse the dimension afterwards?

Reply:
This line was adding the time-dependant coordinate (like time / horizon) if it was not in the weight matrix. I am understanding from your comment that it would already exists in weights if time or horizon was in the datasets?